batcher/batcher.go

```diff
@@ -0,0 +1,266 @@
+package batcher
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/loop/lndclient"
+	"github.com/lightningnetwork/lnd/ticker"
+)
+
+const (
+	// DefaultMaxBatchSize is the maximum number of outputs we'll create in
+	// each batch.
+	DefaultMaxBatchSize = 50
+
+	// DefaultTxPublishTickerInterval is the time between each time we'll
+	// check whether it is time to batch and send the current set of send
+	// requests.
+	DefaultTxPublishTickerInterval = 1 * time.Second
+)
+
+type sendOutputRequest struct {
+	ctx      context.Context
+	txOut    *wire.TxOut
+	cutoff   time.Time
+	respChan chan *wire.MsgTx
+	errChan  chan error
+}
+
+type numPendingOutsRequest struct {
+	ctx      context.Context
+	respChan chan int
+	errChan  chan error
+}
+
+// Config specifies the various configuration options for the Batcher.
+type Config struct {
+	// TxConfTarget is the target confirmation target used for fee
+	// estimation.
+	// TODO(halseth): make per request target?
+	TxConfTarget int32
+
+	// MaxBatchSize is the maximum number of outputs we'll create in each
+	// batch.
+	MaxBatchSize uint32
+
+	// TxPublishTicker is a ticker that should tick at the interval we want
+	// to check whether it is time to batch and send the current set of
+	// send requests.
+	TxPublishTicker ticker.Ticker
+}
+
+// Batcher is used to create on-chain outputs in batch, thereby reducing fees.
+type Batcher struct {
+	cfg                *Config
+	wallet             lndclient.WalletKitClient
+	sendOutputChan     chan *sendOutputRequest
+	numPendingOutsChan chan *numPendingOutsRequest
+}
+
+// New initiates a new Batcher with the given config.
+func New(cfg *Config, walletKit lndclient.WalletKitClient) *Batcher {
+	return &Batcher{
+		cfg:                cfg,
+		wallet:             walletKit,
+		sendOutputChan:     make(chan *sendOutputRequest),
+		numPendingOutsChan: make(chan *numPendingOutsRequest),
+	}
+}
+
+// Run is the main Batcher routine, responsible for handling incoming send
+// requests, collecting them into batches and broadcast. It will run until the
+// passed context is done.
+func (s *Batcher) Run(ctx context.Context) {
+	// sendRequests contain the current set of pending sends we want to batch.
+	var sendRequests []*sendOutputRequest
+
+	// txPublishTicker represents the time between each time we'll check
+	// whether it is time to batch and send the current set of send
+	// requests.
+	s.cfg.TxPublishTicker.Resume()
+	defer s.cfg.TxPublishTicker.Stop()
+
+	for {
+
+		select {
+		// A new send is being requested. Add it to rou set of requests
+		// and start the ticker.
+		case req := <-s.sendOutputChan:
+			// Add the new request to our slice, and sort it.
+			sendRequests = append(sendRequests, req)
```

Comment:
Instead of letting this grow beyond max size, shouldn't we publish right away when we've added the max'th item?